ansible/playbook-setup-kiwi3.yml

```diff
@@ -210,11 +211,40 @@
     changed_when: "docker_network.stdout == '[]\n'"
 
   - name: Launch Docker containers (may take a while!)
-    docker_compose:
-      project_src: "/home/{{ item }}/git-clone"
-      state: present
-      restarted: "{{ item == 'ssl-reverse-proxy' and redirect.changed }}"
-      build: yes
+    shell:
+      chdir: "/home/{{ item }}/git-clone"
+      cmd: |
+        set -e -u -x -o pipefail
+        changed=true
+        docker compose pull
+        # For the upcoming block, the ideas are:
+        # - Check the SHA256s from the build output against all previously existing
+        #   images: if not all of them are matched, we have at least one new image.
+        # - Use the fact that "grep -v" returns 1 when it filtered all lines away.
+        # - Use the fact that "docker compose up" seems to pick up more recent images
+        #   by default (unless --no-recreate is used), see:
+        #   https://docs.docker.com/reference/cli/docker/compose/up/
+        docker compose build \
+            | { grep -F 'writing image' || true ; } \
+            | tee /dev/stderr \
+            | { grep -oE 'sha256:[a-f0-9]{64}' || true ; } \
+            | grep -v -q -F -f <(docker images --quiet --no-trunc | sort | tee /dev/stderr) || changed=false
+        docker_compose_up_args=(
+            # careful with --remove-orphans!
+            --detach
+            --wait
+        )
+        if [[ "{{ item == 'ssl-reverse-proxy' and redirect.changed }}" = True ]]; then
+            docker_compose_up_args+=( --force-recreate )
+            changed=true
+        fi
+        docker compose up "${docker_compose_up_args[@]}"
+        echo
+        echo "changed=${changed}"
```

Comment:
That's a lot of lines of hand-crafted code (and relying on specific strings in the output of `docker compose` looks somewhat brittle).

Ansible's [community.docker.docker_compose_v2](https://docs.ansible.com/ansible/latest/collections/community/docker/docker_compose_v2_module.html) doesn't work?

Reply:
> That's a lot of lines of hand-crafted code (and relying on specific strings in the output of `docker compose` looks somewhat brittle).

@ulm regarding the output checking: if this breaks, the worst that will happen is that Ansible will say that nothing changed while it did, the rest should keep working.

> Ansible's [community.docker.docker_compose_v2](https://docs.ansible.com/ansible/latest/collections/community/docker/docker_compose_v2_module.html) doesn't work?

Not reliably, no.

The symptom is that oftentimes it doesn't even build the new image and I need to SSH in, build manually, and restart containers.